### ISSUE TITLE
factorizing posterior samplers + inclusion of slice sampler

### ISSUE THREAD

**Issue by rattoma** [DynareTeam/dynare#1065]
I am going to make a pull request from by personal branch 
https://github.com/rattoma/dynare/tree/slice

the aim is to add a new posterior sampler, SLICE, but also to make it easier to add further new ones.
There is a bunch of common codes in random walk, independent, TARB, etc. and I did not want to follow that route.

So I wrote, adapting from random walk algorithm, a new general set of routines 
`posterior_sampler_*`
that factorizes the iteration of individual samplers, but the rest of the engine is common: initialization, parallelization, bars, `load_mh_file` etc...

The current implementation fully integrates:
- random walk MH
- independent MH
- slice

It also contains a quick fix for TARB but the latter could be easily fully integrated in the new framework.

An open issue concerns adaptive_metropolis_hastings.m [last commit in 2013], which cannot be easily integrated, so I trapped it in dynare_estimation_1.m keeping the old framework.

Concerning pre-processor provisions, this commit would require:
1) enable slice as new sampler method:
`options_.posterior_sampling_method = 'slice';`
2) enable a form of flexible method specific options, along the lines of optimizers options. would this be possible? To explain better:
I added a new option field `options_.posterior_sampler_options` that can contain method specific options. one could specify a syntax like 
`sampler_options=(NAME, VALUE, etc..)` 

with name/value pairs. These will be stored as fields of 
 `options_.posterior_sampler_options`

So the command
`estimation(posterior_sampling_method = slice, sampler_options=('rotated',1), ...)` 

would trigger

 `options_.posterior_sampler_options.rotated = 1`

what do you think?

**Comment by rattoma** [comment]
the current status of my branch can already be tested

**Comment by stepan-a** [comment]
Thanks @rattoma I will grab your branch and run the tests this week-end.

**Comment by rattoma** [comment]
Thanks @stepan-a. Please note I just committed few changes [should only apply to slice but might affect also other things]

**Comment by stepan-a** [comment]
@rattoma Here are the results of the tests (with matlab) considering the ```slice```branch (after the merge with ```master```, ie commit d1809152107ad0c95b9afe800113d2bec0d207e1):
```
================================
DYNARE MAKE CHECK MATLAB RESULTS
================================
| TOTAL: 561
|  PASS: 507
|  FAIL: 49
| XFAIL: 5
| XPASS: 0
| LIST OF FAILED TESTS:
|     * estimation/fs2000.mod
|     * estimation/fs2000_MCMC_jumping_covariance.mod
|     * estimation/fs2000_initialize_from_calib.mod
|     * estimation/fs2000_model_comparison.mod
|     * estimation/MH_recover/fs2000_recover.mod
|     * estimation/t_proposal/fs2000_student.mod
|     * estimation/TaRB/fs2000_tarb.mod
|     * gsa/ls2003.mod
|     * gsa/cod_ML_morris/cod_ML_morris.mod
|     * gsa/morris/nk_est.mod
|     * discretionary_policy/Gali_discretion.mod
|     * arima/mod1a.mod
|     * arima/mod1c.mod
|     * data/mod1a.mod
|     * fs2000/fs2000.mod
|     * fs2000/fs2000a.mod
|     * fs2000/fs2000_analytic_derivation.mod
|     * fs2000/fs2000_missing_data.mod
|     * fs2000/fs2000_dseries_a.mod
|     * fs2000/fs2000_dseries_b.mod
|     * bvar_a_la_sims/bvar_and_dsge.mod
|     * dsge-var/dsgevar_forward_calibrated_lambda.mod
|     * dsge-var/dsgevar_forward_estimated_lambda.mod
|     * identification/kim/kim2.mod
|     * identification/as2007/as2007.mod
|     * identification/ident_unit_root/ident_unit_root.mod
|     * identification/rbc_ident/rbc_ident_std_as_structural_par.mod
|     * identification/rbc_ident/rbc_ident_varexo_only.mod
|     * conditional_forecasts/2/fs2000_est.mod
|     * recursive/ls2003_bayesian.mod
|     * recursive/ls2003_bayesian_xls.mod
|     * recursive/ls2003_bayesian_csv.mod
|     * kalman_filter_smoother/check_variable_dimensions/fs2000.mod
|     * kalman/likelihood_from_dynare/fs2000_corr_ME.mod
|     * kalman/likelihood_from_dynare/fs2000_corr_ME_missing.mod
|     * kalman/likelihood_from_dynare/fs2000_uncorr_ME.mod
|     * kalman/likelihood_from_dynare/fs2000_uncorr_ME_missing.mod
|     * ep/rbc2.mod
|     * measurement_errors/fs2000_corr_me_ml_mcmc/fs2000_corr_ME.mod
|     * forecast/ls2003_rolling_window_forecast.mod
|     * filter_step_ahead/fs2000_filter_step_ahead_bayesian.mod
|     * smoother2histval/fs2000_smooth.mod
|     * smoother2histval/fs2000_smooth_stoch_simul.mod
|     * optimizers/fs2000_6.mod
|     * TeX/fs2000_corr_ME.mod
|     * prior_posterior_function/fs2000_prior_posterior_function.mod
|     * ../matlab/modules/dseries/src/@dseries/subsasgn.m#21
|     * ../matlab/modules/dseries/src/@dseries/subsasgn.m#22
|     * ../matlab/prior_draw.m#1
|
| LIST OF 10 SLOWEST TESTS:
|     * optimizers/fs2000_2.mod - 712.260000
|     * optimizers/fs2000_6.mod - 614.060000
|     * walsh.mod - 468.390000
|     * estimation/TaRB/fs2000_tarb.mod - 404.430000
|     * gsa/ls2003.mod - 379.840000
|     * optimizers/fs2000_10.mod - 287.400000
|     * ms-sbvar/test_ms_variances_repeated_runs.mod - 287.860000
|     * conditional_forecasts/2/fs2000_est.mod - 278.860000
|     * gsa/ls2003a.mod - 261.910000
|     * optimizers/fs2000_9.mod - 235.990000
```

**Comment by rattoma** [comment]
@stepan-a thank you! 
I will have a look. My first guess is around your rolling back of CutSample after MCMC diagnostics. Computing inefficiency factor needs info in record from CutSample which is most probably now missing. Also, I do not think CutSample should alter the behavior of Brooks/Gelman or Geweke tests [just adds more fields to the record] if we run it before calling McMCDiagnostics.m. Or I am missing something?

**Comment by rattoma** [comment]
@stepan-a. Could you please retry the testsuite on the current state of the branch [i.e. after commits on Oct 14th 2015]?
If things look better I can work on this branch to allow a pull request.

**Comment by stepan-a** [comment]
@rattoma It looks much better!

```
================================
DYNARE MAKE CHECK MATLAB RESULTS
================================
| TOTAL: 561
|  PASS: 536
|  FAIL: 20
| XFAIL: 5
| XPASS: 0
| LIST OF FAILED TESTS:
|     * estimation/MH_recover/fs2000_recover.mod
|     * gsa/ls2003.mod
|     * gsa/cod_ML_morris/cod_ML_morris.mod
|     * discretionary_policy/Gali_discretion.mod
|     * identification/kim/kim2.mod
|     * identification/as2007/as2007.mod
|     * identification/ident_unit_root/ident_unit_root.mod
|     * identification/rbc_ident/rbc_ident_std_as_structural_par.mod
|     * identification/rbc_ident/rbc_ident_varexo_only.mod
|     * kalman/likelihood_from_dynare/fs2000_corr_ME.mod
|     * kalman/likelihood_from_dynare/fs2000_corr_ME_missing.mod
|     * kalman/likelihood_from_dynare/fs2000_uncorr_ME.mod
|     * kalman/likelihood_from_dynare/fs2000_uncorr_ME_missing.mod
|     * ep/rbc2.mod
|     * measurement_errors/fs2000_corr_me_ml_mcmc/fs2000_corr_ME.mod
|     * TeX/fs2000_corr_ME.mod
|     * run_reporting_test_matlab.m
|     * ../matlab/modules/dseries/src/@dseries/subsasgn.m#21
|     * ../matlab/modules/dseries/src/@dseries/subsasgn.m#22
|     * ../matlab/prior_draw.m#1
|
| LIST OF 10 SLOWEST TESTS:
|     * walsh.mod - 858.160000
|     * optimizers/fs2000_2.mod - 771.970000
|     * optimizers/fs2000_6.mod - 685.070000
|     * gsa/ls2003.mod - 568.480000
|     * estimation/TaRB/fs2000_tarb.mod - 528.290000
|     * conditional_forecasts/2/fs2000_est.mod - 403.980000
|     * optimizers/fs2000_10.mod - 334.140000
|     * ms-sbvar/test_ms_variances_repeated_runs.mod - 329.080000
|     * gsa/ls2003a.mod - 326.410000
|     * optimizers/fs2000_9.mod - 286.540000
```
I have the impression that the failing tests are caused by bugs fixed since mid-October. I master and run the testsuite again...

**Comment by rattoma** [comment]
@stepan-a excellent. I think that properly mastering will require merging bug fixes done for

metropolis_hastings_initialization.m
random_walk_metropolis_hastings.m
random_walk_metropolis_hastings_core.m

on the new routines:

matlab\check_posterior_sampler_options.m
matlab\posterior_sampler.m
matlab\posterior_sampler_core.m
matlab\posterior_sampler_initialization.m
matlab\posterior_sampler_iteration.m

which replace the former ones ...

**Comment by rattoma** [comment]
I will work on that

**Comment by rattoma** [comment]
also TaRB il left out: I should factor it out together will all other samplers

**Comment by rattoma** [comment]
@stepan-a would you prefer I rebase, so that slice commits are shown after latest commits, or a merge?

**Comment by rattoma** [comment]
I am about to make a pull request for slice which looks okay.
Notes on preprocessor. We would need to add:
1) slice as an optional `posterior_sampling_method`;
2) allow posterior sampler specific options, similar to `optim` for maximizers. Currently I am using `optim` to get the behavior [as exemplified in the specific test routine I am adding], but it would be good to have something like
`sampler(OPTION, VALUE)` 
that triggers an `options_` field:
`options_.sampler_options`

**Comment by rattoma** [comment]
@stepan-a I have just did a forced push on my personal slice branch [after rebase]. Could you please re-sun the testsuite on it?

**Comment by stepan-a** [comment]
@rattoma Okay I'll do that asap

**Comment by stepan-a** [comment]
@rattoma The output of the testsuite seems ok

```
================================
DYNARE MAKE CHECK MATLAB RESULTS
================================
| TOTAL: 677
|  PASS: 668
|  FAIL: 4
| XFAIL: 5
| XPASS: 0
| LIST OF FAILED TESTS:
|     * kalman/likelihood_from_dynare/fs2000_corr_ME.mod
|     * kalman/likelihood_from_dynare/fs2000_corr_ME_missing.mod
|     * kalman/likelihood_from_dynare/fs2000_uncorr_ME.mod
|     * kalman/likelihood_from_dynare/fs2000_uncorr_ME_missing.mod
|
| LIST OF 10 SLOWEST TESTS:
|     * gsa/ls2003.mod - 1248.090000
|     * optimizers/fs2000_6.mod - 1076.860000
|     * observation_trends_and_prefiltering/MCMC/Trend_loglinear_no_prefilter_MC.mod - 776.520000
|     * observation_trends_and_prefiltering/MCMC/Trend_no_prefilter_MC.mod - 772.450000
|     * observation_trends_and_prefiltering/MCMC/Trend_loglinear_prefilter_MC.mod - 764.430000
|     * estimation/fs2000_slice.mod - 758.750000
|     * observation_trends_and_prefiltering/MCMC/Trend_prefilter_MC.mod - 750.420000
|     * observation_trends_and_prefiltering/MCMC/Trend_loglin_no_prefilt_first_obs_MC.mod - 741.910000
|     * observation_trends_and_prefiltering/MCMC/Trend_loglin_prefilt_first_obs_MC.mod - 726.040000
|     * observation_trends_and_prefiltering/MCMC/Trend_no_prefilter_first_obs_MC.mod - 722.590000
```
I need to run again the testsuite on master to check the timing also...

**Comment by stepan-a** [comment]
Timing does not seem to differ significantly.

**Comment by rattoma** [comment]
excellent thanks a lot

**Comment by stepan-a** [comment]
@rattoma Did you add an example in your branch (in ```./tests```) for the slice sampler?

**Comment by stepan-a** [comment]
Sorry I found it... It was in the last commit.

**Comment by rattoma** [comment]
as usual in my tests it also ranks high in timing ... :(

**Comment by JohannesPfeifer** [comment]
This ticket is closed by #1188
